dali/kernels/imgproc/color_manipulation/brightness_contrast_test.cc

```diff
@@ -0,0 +1,221 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+#include <vector>
+#include <tuple>
+#include "dali/kernels/imgproc/color_manipulation/brightness_contrast.h"
+#include "dali/kernels/test/tensor_test_utils.h"
+
+namespace dali {
+namespace kernels {
+namespace test {
+
+namespace brightness_contrast {
+
+template <template <typename A, typename B> class Pair, typename T1, typename... T>
+using FixedFirstTypePairs = std::tuple<Pair<T1, T>...>;
+
+template <template <typename A, typename B> class Pair, typename A, typename B>
+struct AllPairsHelper;
+
+template <template <typename A, typename B> class Pair, typename... A, typename... B>
+struct AllPairsHelper<Pair, std::tuple<A...>, std::tuple<B...>> {
+  using type = dali::detail::tuple_cat_t<FixedFirstTypePairs<Pair, A, B...>...>;
+};
+
+template <template <typename A, typename B> class Pair, typename TupleA, typename TupleB>
+using AllPairs = typename AllPairsHelper<Pair, TupleA, TupleB>::type;
+
+template <typename Tuple>
+struct TupleToGTest;
+
+template <typename... T>
+struct TupleToGTest<std::tuple<T...>> {
+  using type = testing::Types<T...>;
+};
+
+
+/**
+ * Creates cv::Mat based on provided arguments.
+ * This mat is for roi-testing purposes only. Particularly, it doesn't care
+ * about image type (i.e. number of channels), so don't try to imshow it.
+ *
+ * @param ptr Can't be const, due to cv::Mat API
+ * @param rows height of the input image
+ * @param cols width of the input image
+ */
+template <int nchannels, class T, class Roi>
+cv::Mat_<T> to_mat(T *ptr, Roi roi, int rows, int cols) {
+  auto roi_w = roi.extent().x;
```

Comment:
This one works well:
```
template <int nchannels, class T, class Roi>
cv::Mat_<cv::Vec<T, nchannels>> to_mat(const T *ptr, Roi roi, int rows, int cols) {
  auto roi_w = roi.extent().x;
  auto roi_h = roi.extent().y;
  cv::Mat mat(rows, cols,
              CV_MAKETYPE(cv::DataDepth<std::remove_const_t<T>>::value, nchannels),
              const_cast<T*>(ptr));
  cv::Rect rect(roi.lo.x, roi.lo.y, roi_w, roi_h);
  cv::Mat_<cv::Vec<T, nchannels>> out_copy;
  mat(rect).copyTo(out_copy);
  return out_copy;
}
```

Reply:
In general, I don't like the way your function alters the meaning of Roi dimensions. That kind of behavior is OK internally, but to return a matrix with combined width/channels seems a bit excessive.